### PR TITLE
Crashing is always caught by our handler

### PR DESCRIPTION
**Pull request by bkase**
Moved the monitor above the coda initialization so we can always catch
things.

Tested by throwing an exception asynchronously before coda
initialization finishes.

Later: Handle the exception better and print a nice message

### Reviewers
* Please review with "No Whitespace" checked
* I acknowledge we should refactor this code; however, this file is in need of a bigger rewrite to be on top of graphql directly, so I don't want to invest in cleaning this us now

### CODE REVIEW

File: src/app/cli/src/coda.ml [hunk 1]
```diff
@@ -133,284 +133,312 @@ let daemon logger =
...
+       let monitor = Async.Monitor.create ~name:"coda" () in
+       let module Coda_initialization = struct
+         type ('a, 'b, 'c, 'd, 'e) t =
+           { coda: 'a
+           ; client_whitelist: 'b
+           ; rest_server_port: 'c
+           ; client_port: 'd
+           ; run_snark_worker_action: 'e }
+       end in
+       let coda_initialization_deferred =
```
Comment (cmr) [inline src/app/cli/src/coda.ml]: This doesn't prevent the code inside the deferred from running. You should make this a `let coda_initialized_deferred () = ` to have the effect you want.

File: src/app/cli/src/coda.ml [hunk 2]
```diff
@@ -133,284 +133,312 @@ let daemon logger =
...
+         { Coda_initialization.coda
+         ; client_whitelist
+         ; rest_server_port
+         ; client_port
+         ; run_snark_worker_action }
        in
-       Coda_run.handle_shutdown ~monitor ~conf_dir coda ;
+       Coda_run.handle_shutdown ~monitor ~conf_dir ~top_logger:logger
+         (coda_initialization_deferred >>| fun c -> c.Coda_initialization.coda) ;
+       let crash () = raise TestException in
```
Comment (cmr) [inline src/app/cli/src/coda.ml]: This `raise` is leftover, right?

File: src/app/cli/src/coda.ml (hunk 1 above)
Comment (bkase) [reply src/app/cli/src/coda.ml]: Actually I'm immediately starting the deferred there on purpose, but perhaps its cleaner to start it closer (in the sense of the line of code) to where it's being used

Comment (cmr) [reply src/app/cli/src/coda.ml]: By starting it here, you aren't running it inside the monitor.

File: src/app/cli/src/coda_worker.ml [hunk 3]
```diff
@@ -425,7 +425,9 @@ module T = struct
                  ~initial_propose_keypairs ~monitor ~consensus_local_state
                  ~transaction_database ~external_transition_database ())
           in
-          Coda_run.handle_shutdown ~monitor ~conf_dir coda ;
+          Coda_run.handle_shutdown ~monitor ~conf_dir ~top_logger:logger
+            coda_deferred ;
+          let%bind coda = coda_deferred in
```
Comment (cmr) [inline src/app/cli/src/coda_worker.ml]: This isn't running in one of our monitors.

File: src/app/cli/src/coda.ml [hunk 4]
```diff
@@ -133,284 +133,312 @@ let daemon logger =
...
+                ~transaction_database ~external_transition_database ())
+         in
+         { Coda_initialization.coda
+         ; client_whitelist
+         ; rest_server_port
+         ; client_port
+         ; run_snark_worker_action }
        in
-       Coda_run.handle_shutdown ~monitor ~conf_dir coda ;
+       let coda_initialization_deferred = coda_initialization_deferred () in
```
Comment (cmr) [inline src/app/cli/src/coda.ml]: This still isn't running inside our monitor.

Comment (bkase) [reply src/app/cli/src/coda.ml]: but at least it's captured at the first bind-point (and anecdotally that is sufficient for capturing the exceptions during this period)

**Pull request merged**